### PR TITLE
Use JSON visitor for serialization

### LINKED ISSUE #423: [FEATURE] Use visitor pattern for the JSON serializer and deserializer
In #420 we have used visitor pattern for the deserialization of `msgpack`. We can use similar pattern for  `json` serialization. 

# Serializer
We can use `msgpack` visitor pattern for the `msgpack` to `json` conversion in the serializer. There will be two benefits:

1. Skip the intermediate data struct from `nlohmann_json`.
2. Allow partial indent `json` output, which is preferred for readability. Such as below.

```json
{
  "data": {
    "node": [
      {"id": 5, "u_rated": 10e3},
      {"id": 6, "u_rated": 150e3}
    ]
  } 
}
```

# Deserializer

We can use [SAX API](https://json.nlohmann.me/features/parsing/sax_interface/) of other `json` libraries for `json` to `msgpack` conversion.

### PR DESCRIPTION
**Pull request by TonyXiang8787**
Closes #423

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp [hunk 1]
```diff
@@ -59,6 +63,147 @@ MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS) {
...
+        return true;
+    }
+    bool visit_float32(float v) { return visit_float64(v); }
+    bool visit_float64(double v) {
+        if (std::isinf(v)) {
+            if (v > 0.0) {
+                ss << '"' << "inf" << '"';
+            } else {
+                ss << '"' << "-inf" << '"';
+            }
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp:138]: why not just `'"inf"'`?
```suggestion
            if (v > 0.0) {
                ss << '"inf"';
            } else {
                ss << '"-inf"';
            }
```

if anything, i would just do

```suggestion
            ss << '"' << (v > 0.0 ? "" : "-") << "inf" << '"';
```

or even

```suggestion
            ss << (v > 0.0 ? '"inf"' : '"-inf"');
```

or

```suggestion
            auto const str = v > 0.0 ? "inf"sv: "-inf"sv;
            visit_str(str.c_str(), str.size());
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp [hunk 2]
```diff
@@ -59,6 +63,147 @@ MSGPACK_API_VERSION_NAMESPACE(MSGPACK_DEFAULT_API_NS) {
...
+                return;
+            }
+            ss << sep_char;
+            return;
+        }
+        ss << '\n';
+        Idx indent_size = indent_level * indent;
+        while (indent_size-- != 0) {
+            ss << sep_char;
+        }
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/serializer.hpp:101]: maybe this is more readable (dw, i'll try before committing)

```suggestion
        ss << std::string{.count = indent_level * indent, .ch = sep_char};
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 3]
```diff
@@ -38,6 +39,101 @@ constexpr from_json_t from_json;
...
+    }
+    static bool binary(json::binary_t const& /* val */) { return true; }
+
+    bool start_object(size_t /* elements */) {
+        data_buffers.emplace();
+        return true;
+    }
+    bool end_object() {
+        JsonMapArrayData object_data{std::move(data_buffers.top())};
+        data_buffers.pop();
```
Comment (mgovers) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:94]: double-checking but i think if you `pop`, you get the value as well, right? (will check myself)

File: tests/cpp_unit_tests/test_serializer.cpp [hunk 4]
```diff
@@ -6,19 +6,156 @@
...
+    R"({"version":"1.0","type":"update","is_batch":true,"attributes":{},"data":[{"sym_load":[{"id":9,"p_specified":10},{"id":10}],"asym_load":[{"id":5,"p_specified":[10,11,12]},{"id":6,"p_specified":[15,null,16]}]},{"sym_load":[{"id":11,"p_specified":"inf"},{"id":12,"p_specified":"-inf"}],"asym_load":[{"id":13,"p_specified":["inf",11,17]},{"id":14,"p_specified":[10,"-inf",19]}],"asym_gen":[{"id":7}]}]})";
 constexpr std::string_view batch_dataset_dict =
-    R"({"attributes":{"asym_gen":["id"],"asym_load":["id","p_specified"],"sym_load":["id","p_specified"]},"data":[{"asym_load":[[5,[10,11,12]],[6,[15,null,16]]],"sym_load":[[9,10],[10,null]]},{"asym_gen":[[7]],"asym_load":[[13,["inf",11,17]],[14,[10,"-inf",19]]],"sym_load":[[11,"inf"],[12,"-inf"]]}],"is_batch":true,"type":"update","version":"1.0"})";
+    R"({"version":"1.0","type":"update","is_batch":true,"attributes":{"asym_gen":["id"],"asym_load":["id","p_specified"],"sym_load":["id","p_specified"]},"data":[{"sym_load":[[9,10],[10,null]],"asym_load":[[5,[10,11,12]],[6,[15,null,16]]]},{"sym_load":[[11,"inf"],[12,"-inf"]],"asym_load":[[13,["inf",11,17]],[14,[10,"-inf",19]]],"asym_gen":[[7]]}]})";
+} // namespace
+
+namespace {
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_serializer.cpp:25]:
```suggestion

```

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp (hunk 3 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp:94]: nope, `pop` is `void`

File: power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp [hunk 5]
```diff
@@ -44,18 +44,21 @@ using nlohmann::json;
 // visitor for json conversion
 struct JsonMapArrayData {
     size_t size{};
-    msgpack::sbuffer buffer{};
```
Comment (TonyXiang8787) [inline power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp]: why use a `shared_ptr` here? It does not make any sense?

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp]: `msgpack::sbuffer` has copy constructor `delete` but the copy constructor is a prerequisite for `std::stack`. This makes it copyable. If you have any better ideas please lmk

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp]: cfr discussion: reverted and we'll see what sonar cloud thinks of it

**Review by mgovers — APPROVED** [review]

**Pull request merged**